metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,186 @@
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Boolean] = {
+    {
+      //TODO: Validate transactions PM-3131/3132
+      false.pure[F]
+    } >>
+      ledgerUpdateSemaphore.withPermit {
+        updateLedger(block).map {
+          _.hash == block.header.postStateHash
+        }
+      }
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Unit] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+    if (commitQC.blockHash != block.hash) {
+      //Several options here, depending on how we want to schedule the execution (BlockExecutor):
+      // - disallow this condition
+      // - ignore
+      // - execute block without constructing CC
+      ().pure[F]
+    } else {
+      for {
+        _          <- ledgerUpdateSemaphore.acquireN(maxConcurrentUpdates)
+        ledger     <- updateLedger(block)
+        checkpoint <- constructCheckpoint(ledger, commitQC)
+        _          <- saveCommittedLedger(ledger)
+        _          <- resetLedgerTree(ledger, block)
+        _          <- ledgerUpdateSemaphore.releaseN(maxConcurrentUpdates)
+        _          <- pushCheckpoint(checkpoint)
+      } yield ()
+    }
+  }
+
+  private def updateLedger(block: Block): F[Ledger] = {
+
+    // I'm worried that this may loop for a long time (until it reaches a pruned block)
+    // when the initial block is an ancestor of the last committed block - could that happen?
+    def loop(block: Block): F[Ledger] = {
+
+      getLedger(block.header.parentHash).flatMap {
+        case Some(ledger) =>
+          updateLedgerByBlock(ledger, block)
+
+        case None =>
+          getBlock(block.header.parentHash).flatMap {
+            case Some(parent) =>
+              loop(parent) >>= (updateLedgerByBlock(_, block))
+
+            case None =>
+              Sync[F].raiseError(
+                new IllegalStateException(s"Block not available: ${block.hash}")
+              )
+          }
+      }
+
+    }
+
+    OptionT(getLedger(block.hash)).getOrElseF(loop(block))
+  }
+
+  private def updateLedgerByBlock(
+      oldLedger: Ledger,
+      block: Block
+  ): F[Ledger] = {
+    val newLedger = block.body.transactions.foldLeft(oldLedger)(_ update _)
+    ledgerTree
+      .getAndUpdate {
+        _ + (block.hash -> (newLedger, block.header))
+      }
+      .as(newLedger)
+  }
+
+  private def getBlock(hash: Block.Hash): F[Option[Block]] =
+    storeRunner.runReadOnly(blockStorage.get(hash))
+
+  private def getLedger(hash: Block.Hash): F[Option[Ledger]] =
+    ledgerTree.get.map(_.get(hash).map(_._1))
+
+  private def saveCommittedLedger(ledger: Ledger): F[Unit] =
+    storeRunner.runReadWrite {
+      ledgerStorage.put(ledger)
+    }
+
+  private def resetLedgerTree(ledger: Ledger, block: Block): F[Unit] =
+    ledgerTree.set(Map(block.hash -> (ledger, block.header)))
```

Comment:
Hm, this seems easy :) I haven't thought about it, always thought in terms of pruning based on ancestry. But it seems true that the commit will always belong to a tip.

Reply:
> Not a problem, just wanted to bring it up that when we see a Commit QC here, it may not be the tip any more, thanks to concurrency.

If we will receive a Commit QC for an ancestor of the last executed block then the `updateLedger` method should raise an error, now that it's looking to block heights

Reply:
True, what I meant is that you may be discarding states from the tree that you will still need. 

Say we had this chain where `*` is the first, `**` is the second and `***` is the third commit. When the executor is asked to execute blocks `[C,D]`, it might happen that while `C` is executing, we have already received a commit for `E` and you are being asked to validate `F`. So you construct the ledger for `D`, `E` and `F` in the tree, and _then_ finally get to actually execute `D` because `C` was slow.
```
A - B* - C - D** - E*** - F
```
At that point your code will clear out all content from the tree apart from the ledger that belonged to `D`, and when it comes to execute `E` it will reconstruct it again. 

That's what I was trying to say with the descendant stuff, that if we extracted a generic `KVTreeStorage` then it could say after execution to "prune anything that doesn't descend from D" and keep D, E and F.

Reply:
To put my money where my mouth is I extracted a `KVTree` in https://github.com/input-output-hk/metronome/pull/56

Up to you if you want to use it or not (with an in-memory store, like the `BlockSynchronizer`), but at least it's available.

Reply:
What would be the benefit of using `KVTree` instead?

Reply:
Well, you have a `ledgerTree`, which is a map of ledger and header indexed by block hash; while you can traverse upwards because the headers show you the parent, it's not easy to go in the other direction, towards the descendants. As you noted in tests, `executeBlock` removes all other entries from the tree, potentially causing re-evaluation of the update rules, repeated block retrievals. You identified as a possible optimisation that these states could be kept in the map. But that requires more coding to select which blocks are descendants. 

This is where the `KVTree` could help, since it already has methods to implement this kind of pruning, so you don't have to do it again. It's still not a panacea from the concurrency perspective, as I myself had to be careful while maintaining the in-memory block tree in `BlockSynchronizer` to make sure I don't end up with a forest. That's why I'm not trying to push it on you, just wanted to say that if that optimisation was troublesome, this could be a way to reuse what we already have.

Reply:
I decided against using `KVTree` for 2 reasons:
- setup complexity
- maintaining parent and child links would be difficult to get right in the context of parallel updates

In the current design we either `clearLedgerTree` or only add entries to it. This can be considered thread-safe in our application. By taking advantage of `height` and parent links, pruning the `Map` of non-descendants wasn't too difficult and should be reasonably efficient

Reply:
Sweet, glad we explored all options :+1: